File: exercises/cpp/lab11/overlay.cpp
// Lab 11 - Process Overlay using execl()
// Demonstrates overlaying one process image with another.
// Compile:  g++ -o hello hello.cpp && g++ -o loop loop.cpp && g++ -o overlay overlay.cpp
// Run:      ./overlay
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/wait.h>

int main() {
    printf("=== Process Overlay Demo ===\n\n");

    pid_t pid = fork();

    if (pid < 0) {
        perror("fork");
        return 1;
    } else if (pid == 0) {
        // Child: overlay with hello
        printf("Child  => Running Hello World program via execl\n");
        execl("./hello", "./hello", (char*)NULL);
        perror("execl hello failed"); // only reached on error
        exit(1);
    } else {
        // Parent: wait, then overlay with loop
        wait(NULL);
        printf("Parent => Running loop program via execl\n");
        execl("./loop", "./loop", "5", (char*)NULL);
        perror("execl loop failed");
    }

    return 0;
}

File: exercises/python/lab11/overlay.py
"""Lab 11 - Process Overlay using os.execl()"""
import os
import sys
import time

def main():
    print("=== Process Overlay Demo ===\n")

    pid = os.fork()

    if pid < 0:
        print("fork failed", file=sys.stderr)
        sys.exit(1)
    elif pid == 0:
        print("Child  => Running hello.py via execl")
        os.execl(sys.executable, sys.executable, os.path.join(os.path.dirname(__file__), "hello.py"))
        print("This won't print")
    else:
        os.wait()
        print("Parent => Running loop.py via execl")
        os.execl(sys.executable, sys.executable,
                 os.path.join(os.path.dirname(__file__), "loop.py"), "5")
        print("This won't print")


if __name__ == "__main__":
    main()
